phpBB/phpbb/install/module/update_database/task/update_extensions.php

```diff
@@ -0,0 +1,201 @@
+
+	/** @var update_helper */
+	protected $update_helper;
+
+	/**
+	 * @var \phpbb\config\db
+	 */
+	protected $config;
+
+	/**
+	 * @var \phpbb\log\log_interface
+	 */
+	protected $log;
+
+	/**
+	 * @var \phpbb\user
+	 */
+	protected $user;
+
+	/** @var \phpbb\extension\manager */
+	protected $extension_manager;
+
+	/** @var Finder */
+	protected $finder;
+
+	/** @var string Extension table */
+	protected $extension_table;
+
+	/** @var \phpbb\db\driver\driver_interface */
+	protected $db;
+
+	/**
+	 * Constructor
+	 *
+	 * @param container_factory			$container
+	 * @param config					$install_config
+	 * @param iohandler_interface		$iohandler
+	 * @param $update_helper			$update_helper
+	 * @param string					$phpbb_root_path phpBB root path
+	 */
+	public function __construct(container_factory $container, config $install_config, iohandler_interface $iohandler, update_helper $update_helper, $phpbb_root_path)
+	{
+		$this->install_config	= $install_config;
+		$this->iohandler		= $iohandler;
+		$this->extension_table = $container->get_parameter('tables.ext');
+
+		$this->log				= $container->get('log');
+		$this->user				= $container->get('user');
+		$this->extension_manager = $container->get('ext.manager');
+		$this->config			= $container->get('config');
+		$this->db				= $container->get('dbal.conn');
+		$this->finder = new Finder();
+		$this->finder->in($phpbb_root_path . 'ext/')
+			->ignoreUnreadableDirs()
+			->depth('< 3')
+			->files()
+			->name('composer.json');
+
+		// Make sure asset version exists in config. Otherwise we might try to
+		// insert the assets_version setting into the database and cause a
+		// duplicate entry error.
+		if (!isset($this->config['assets_version']))
+		{
+			$this->config['assets_version'] = 0;
+		}
+
+		parent::__construct(true);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function run()
+	{
+		$this->user->session_begin();
+		$this->user->setup(array('common', 'acp/common', 'cli'));
+
+		$input = new ArgvInput();
+		$update_extensions = explode(',', $input->getArgument('update-extensions'));
+
+		$update_info = $this->install_config->get('update_info_unprocessed', array());
+
+		if (!empty($update_info))
+		{
+			// Find available extensions
+			foreach ($this->finder as $file)
+			{
+				/** @var \SplFileInfo $file */
+				$ext_name = preg_replace('#(.+[\\/\\\]ext[\\/\\\])(\w+)[\\/\\\](\w+)#', '$2/$3', dirname($file->getRealPath()));
+
+				// Skip extensions that were not added or updated during update
+				if (!count(preg_grep('#ext/' . $ext_name . '#', $update_info['files'])) &&
+					!in_array($ext_name, $update_extensions) && $ext_name !== 'phpbb/viglink')
+				{
+					continue;
+				}
+
+				// Disable enabled extensions in order to run migrations if needed
+				if ($this->extension_manager->is_enabled($ext_name))
+				{
+					$this->extension_manager->disable($ext_name);
+				}
+
+				if ($this->extension_manager->is_available($ext_name))
+				{
+					$this->extension_manager->enable($ext_name);
```

Comment:
that's the difficult part yes. And indeed I think we don't want to enable an extensions which wasn't previously enabled. Except if it is in a list of newly extensions enabled by default (imho, the list should be provided by the update package, like we provide the list of changed files)/